# Report
### Hong Pengfei               
### Gao Yunyi 
### Wu Tianyu
 -------------------------------------------------------------------------

In [5]:
from PGM.GM.HMM.hidden_markov_model import *
from pathlib import Path

#### The input data is stored in List[List[str]] where: <br>
#### for sentences:  each string is one word, each sublist is one sentence<br>
#### for label: each string is one label, each sublist is labels for one sentence in corresponding sequence

### Get data

In [7]:
DATA_FOLDER = Path('./dataset/')
AL = DATA_FOLDER/'AL'
AL_train = AL/'train'
AL_dev_x = AL/'dev.in'
AL_dev_y = AL/'dev.out'
AL_out_2 = AL/'dev.p2.out'
AL_out_3 = AL/'dev.p3.out'
AL_out_4 = AL/'dev.p4.out'
AL_crf_1 = AL/'dev.crf1.out'
Al_crf_2 = AL/'dev.crf2.out'
AL_memm_1 = AL/'dev.memm1.out'
AL_sp_1 = AL/'dev.sp1.out'


EN = DATA_FOLDER/'EN'
EN_train = EN/'train'
EN_dev_x = EN/'dev.in'
EN_dev_y = EN/'dev.out'
EN_out_2 = EN/'dev.p2.out'
EN_out_3 = EN/'dev.p3.out'
EN_out_4 = EN/'dev.p4.out'
EN_crf_1 = EN/'dev.crf1.out'
EN_memm_1 = EN/'dev.memm1.out'
EN_sp_1 = EN/'dev.sp1.out'

CN = DATA_FOLDER/'CN'
CN_train = CN/'train'
CN_dev_x = CN/'dev.in'
CN_dev_y = CN/'dev.out'
CN_out_2 = CN/'dev.p2.out'
CN_out_3 = CN/'dev.p3.out'

SG = DATA_FOLDER/'SG'
SG_train = SG/'train'
SG_dev_x = SG/'dev.in'
SG_dev_y = SG/'dev.out'
SG_out_2 = SG/'dev.p2.out'
SG_out_3 = SG/'dev.p3.out'

PARA_FOLDER = DATA_FOLDER/'parameter'
AL_para_FOLDER = PARA_FOLDER/'AL'
EN_para_FOLDER = PARA_FOLDER/'EN'


EVAL_script = './EvalScript/evalResult.py'

## Part2

#### The count is stored in a $|V|\times|T|$ matrix

#### The emission probabilities are obtained by dividing each entry in the counting matrix by the sum of corresponding row

#### The prediction algorithm is simple: find the maximum entry's column index and convert it to tag

##### AL

In [9]:
hmm = HMM(AL_train)
hmm.train_emission()
hmm.navie_predict(AL_dev_y, AL_out_2)
! python {EVAL_script} {AL_dev_y} {AL_out_2}


#Entity in gold data: 8408
#Entity in prediction: 19484

#Correct Entity : 2898
Entity  precision: 0.1487
Entity  recall: 0.3447
Entity  F: 0.2078

#Correct Sentiment : 2457
Sentiment  precision: 0.1261
Sentiment  recall: 0.2922
Sentiment  F: 0.1762


#### EN

In [10]:
hmm = HMM(EN_train)
hmm.train_emission()
hmm.navie_predict(EN_dev_y, EN_out_2)
! python {EVAL_script} {EN_dev_y} {EN_out_2}


#Entity in gold data: 13179
#Entity in prediction: 19406

#Correct Entity : 9152
Entity  precision: 0.4716
Entity  recall: 0.6944
Entity  F: 0.5617

#Correct Sentiment : 7644
Sentiment  precision: 0.3939
Sentiment  recall: 0.5800
Sentiment  F: 0.4692


#### SG

In [11]:
hmm = HMM(SG_train)
hmm.train_emission()
hmm.navie_predict(SG_dev_y, SG_out_2)
! python {EVAL_script} {SG_dev_y} {SG_out_2}


#Entity in gold data: 4537
#Entity in prediction: 18451

#Correct Entity : 2632
Entity  precision: 0.1426
Entity  recall: 0.5801
Entity  F: 0.2290

#Correct Sentiment : 1239
Sentiment  precision: 0.0672
Sentiment  recall: 0.2731
Sentiment  F: 0.1078


#### CN

In [12]:
hmm = HMM(CN_train)
hmm.train_emission()
hmm.navie_predict(CN_dev_y, CN_out_2)
! python {EVAL_script} {CN_dev_y} {CN_out_2}


#Entity in gold data: 1478
#Entity in prediction: 9373

#Correct Entity : 765
Entity  precision: 0.0816
Entity  recall: 0.5176
Entity  F: 0.1410

#Correct Sentiment : 285
Sentiment  precision: 0.0304
Sentiment  recall: 0.1928
Sentiment  F: 0.0525


## Part 3

####  <SOS\> and <EOS\> are added at the beginning and end of sentence respectively.

#### Traverse through each sentence. 

#### The count is stored in a $(|T|+1)\times(|T|+1)$ matrix. The rows are initial state and columns are the state trasitted to.

#### The probabilities are calculated by dividing each entry with the sum of correspodning row.

#### For Viterbi Algorithm, one parent matrix and one score matrix of size $n\times(|T|+1)$ are implemented

#### The optimal path is restored by finding the argmax among $|T|$ tags in previous position

#### In our implementation, the code snippet is the same as Part 4 since Part 4 addresses a more general problem.

#### To make use of code snippet for Part 4, simply set k=1 for the optimal case

In [13]:
hmm = HMM(AL_train)
hmm.train()
hmm.predict_top_k(AL_dev_y, AL_out_3, k=1)
! python {EVAL_script} {AL_dev_y} {AL_out_3}


#Entity in gold data: 8408
#Entity in prediction: 8475

#Correct Entity : 6717
Entity  precision: 0.7926
Entity  recall: 0.7989
Entity  F: 0.7957

#Correct Sentiment : 6068
Sentiment  precision: 0.7160
Sentiment  recall: 0.7217
Sentiment  F: 0.7188


In [14]:
hmm = HMM(EN_train)
hmm.train()
hmm.predict_top_k(EN_dev_y, EN_out_3, k=1)
! python {EVAL_script} {EN_dev_y} {EN_out_3}


#Entity in gold data: 13179
#Entity in prediction: 12723

#Correct Entity : 10791
Entity  precision: 0.8481
Entity  recall: 0.8188
Entity  F: 0.8332

#Correct Sentiment : 10378
Sentiment  precision: 0.8157
Sentiment  recall: 0.7875
Sentiment  F: 0.8013


In [15]:
hmm = HMM(SG_train)
hmm.train()
hmm.predict_top_k(SG_dev_y, SG_out_3, k=1)
! python {EVAL_script} {SG_dev_y} {SG_out_3}


#Entity in gold data: 4537
#Entity in prediction: 3036

#Correct Entity : 1662
Entity  precision: 0.5474
Entity  recall: 0.3663
Entity  F: 0.4389

#Correct Sentiment : 1035
Sentiment  precision: 0.3409
Sentiment  recall: 0.2281
Sentiment  F: 0.2733


In [16]:
hmm = HMM(CN_train)
hmm.train()
hmm.predict_top_k(CN_dev_y, CN_out_3, k=1)
! python {EVAL_script} {CN_dev_y} {CN_out_3}


#Entity in gold data: 1478
#Entity in prediction: 769

#Correct Entity : 309
Entity  precision: 0.4018
Entity  recall: 0.2091
Entity  F: 0.2750

#Correct Sentiment : 210
Sentiment  precision: 0.2731
Sentiment  recall: 0.1421
Sentiment  F: 0.1869


## Part 4 Top 7th sequences

#### Parent 'matrix' and score 'matrix' are implemented as a numpy 3-d array of size $(n+2)\times(|T|+1)\times K$.

#### The new dimension stores the top K scores and corresponding parent nodes for all positions and tags

#### How to backtrack K-th best path:

#### Repetition begins: select the parent tags for the first K paths for the <EOS\> position and identify the parent tag 'T' corresponding to the K-th best path among all $|T|\times K$ possible paths

#### Count the number of apperances of 'T' in the K parent tags for the first K paths.  Call this number ORDER

#### Repeat the repetitive steps describe above for one position before, this time select first ORDER instead of K

#### Terminate until recovering the entire path

In [17]:
hmm = HMM(AL_train)
hmm.train()
hmm.predict_top_k(AL_dev_y, AL_out_4, k=7)
! python3 {EVAL_script} {AL_dev_y} {AL_out_4}


#Entity in gold data: 8408
#Entity in prediction: 8932

#Correct Entity : 5980
Entity  precision: 0.6695
Entity  recall: 0.7112
Entity  F: 0.6897

#Correct Sentiment : 5014
Sentiment  precision: 0.5614
Sentiment  recall: 0.5963
Sentiment  F: 0.5783


In [18]:
hmm = HMM(EN_train)
hmm.train()
hmm.predict_top_k(EN_dev_y, EN_out_4, k=7)
! python3 {EVAL_script} {EN_dev_y} {EN_out_4}


#Entity in gold data: 13179
#Entity in prediction: 13153

#Correct Entity : 10232
Entity  precision: 0.7779
Entity  recall: 0.7764
Entity  F: 0.7772

#Correct Sentiment : 9719
Sentiment  precision: 0.7389
Sentiment  recall: 0.7375
Sentiment  F: 0.7382


# Part 5  

# model 1 : HMM

### Advantage of HMM:
1. learns distribution between p(x, y), performs generally better when there is an small dataset

### Disadvantage of HMM:

2. HMM is a generative model modelling the distributon of x,y at the same time. When the distribution of x for training dataset is different from the distribution of x for testing dataset (which is very likely), the performance on testing dataset would be very poor.

3. Features of input x is ignored in the model. In NLP tasks, the variety of features in languages is discarded and ignored.

we have found that after adding the smoothing to the parameters, all the performances have increased up by 2~3 percent.

## EN


Entity  F: 0.8596


Sentiment  F: 0.7956

## AL

Entity  F: 0.8157


Sentiment  F: 0.7288

# model 2 : MEMM

### Advatange of MEMM:

1. Discriminative model, solving the issue mentioned in disadvantage of HMM

2. Flexible for different feature functions, mining language features

### Disadvantage of MEMM:

Labelling bias: MEMM sometimes cannot capture the general trend of state transitions because it adopts local variance normalization. The local optimal solution is achieved but it might not be the globally optimal one. The model is limited by this issue, especially when the numbers of convertible states vary significantly among states. Too many convertible states diluted the transition probability, resulting in unequal optimal solution and global solution.

## EN

In [ ]:
EN_para_MEMM_1=EN_para_FOLDER/'MEMM_1'
datafile = EN_train
modelfile = EN_para_MEMM_1
regularization = 10
epoch = 30
memm = DM.get_MEMM(datafile, modelfile, regularization)
memm.load_model(modelfile)
memm.load_data()
memm.train(epoch,lr=0.0.05,start_epoch=4)

#### performance

Entity  F: 0.8330

Sentiment  F: 0.7599

## AL

In [ ]:
AL_para_MEMM_1=AL_para_FOLDER/'MEMM_1'
datafile = AL_train
modelfile = AL_para_MEMM_1
regularization = 10
epoch = 3
memm = DM.get_MEMM(datafile, modelfile, regularization)
memm.load_data()
memm.train(epoch,lr=0.05)

In [ ]:
memm = memm.get_SP(datafile, modelfile, regularization)
memm.load_model(modelfile)
memm.test(AL_dev_y, AL_memm_1)
! python3 {EVAL_script} {AL_dev_y} {AL_memm_1}

Entity  F: 0.7833

Sentiment  F: 0.7299

#### Therefore, we move on from MEMM to CRF, which addresses the labelling bias problem perfectly

# model 3 : CRF

### Advantage of CRF:

1. Compared to HMM: There is no Markov assumptions, which is sometimes unreasonable.

2. Compared to MEMM: Global variance normalization is adopted. Global trend is captured. Labelling bias problem is addressed.

3. Feature functions are flexible, same as MEMM

### Disadvantage of CRF:

1. Computationally complicated. It takes our model hours and hours to train. This also results in low flexibility when new training data is availble.

## Implementation details

In the implementation, we have tried: 
1. different Regularization constants sigma square to prevent overfiting for CRF, we found 1/10 ~ 1/20 to be a good range.

2. different feature functions to extract the most information out of the word. for example, we have tried encode the capitalization of the first word as a feature and lower case everything. we also tried encode numbers into a fixed symbol NUM to deal with every number as a key word. and we use a vocabulary of the words appear more than two times, all of which helped as an extra function.

3. we also tried to smooth the weights by adding a small constant to the weights, and we found that can be useful.

4. we have also tried different window sizes of 1, 2, 3, 4, 5 to encode adjcent word information. we found that a window size of 2 is of great help.



however, due to the limitation, we are not able to fully exploit the search space for all the hyperparameter matching.



training implementation details:
1. we used everything in log space to prevent overflow problems. specifically we also use a special trick in mathematics to deal with overflow problems in calculating log-sum-exp function
2. to replace the lbfgs function in scipy, we have tried to train using gradient descent, we have tried differnet methods to quickly update the parameters. allowing learning rate to decay with the number of epochs increase. 
3. we also tried implementing CRF batching training to save time. but did not manage to properly write the code.
4. we tried using a small number of samples to update the learning rate, but it did not help as well.

## EN

In [ ]:
EN_para_CRF_1=EN_para_FOLDER/'CRF_1'
datafile = EN_train
modelfile = EN_para_CRF_1
regularization = 10
epoch = 10
crf = DM.get_CRF(datafile, modelfile, regularization)
crf.load_model(modelfile)
crf.load_data()
crf.train(epoch,lr=0.01)

In [ ]:
crf = DM.get_CRF(datafile, modelfile, regularization)
crf.load_model(modelfile)
crf.test(EN_dev_y,EN_crf_1)
! python3 {EVAL_script} {EN_dev_y} {EN_crf_1}

#### performance

Entity  F: 0.9094

Sentiment  F: 0.88884

## AL

In [ ]:
AL_para_CRF_2=AL_para_FOLDER/'CRF_2'
datafile = AL_train
modelfile = AL_para_CRF_2
regularization = 10
epoch = 30
crf = DM.get_CRF(datafile, modelfile, regularization)
crf.load_data()
crf.train(epoch,lr=0.1)

#### loss

In [ ]:
AL_crf_2 = AL/'dev.crf2.out'
crf = DM.get_CRF(datafile, modelfile, regularization)
crf.load_model(modelfile)
crf.test(AL_dev_y, AL_crf_2)
! python3 {EVAL_script} {AL_dev_y} {AL_crf_2}

#### performance

# model 4 : SP

we also explored Structural perceptron algorithm since it is just changing the sum operation into max. although we have gotten comparable results with CRF, this approach is faster than CRF by a small amount of time.

we applied CRF as our final submission because the result is stable and will always converge to a stable value. whereas SP may not converge to the small amount of loss every time we run it.

## EN

In [ ]:
EN_para_SP_1=EN_para_FOLDER/'SP_1'
datafile = EN_train
modelfile = EN_para_SP_1
regularization = 10
epoch = 30
sp = DM.get_SP(datafile, modelfile, regularization)
sp.load_data()
sp.train(epoch,lr=)

## AL

In [ ]:
AL_para_SP_1=AL_para_FOLDER/'SP_1'
datafile = AL_train
modelfile = AL_para_SP_1
regularization = 10
epoch = 30
sp = DM.get_SP(datafile, modelfile, regularization)
sp.load_data()
sp.train(epoch,lr=0.01)

#### performance

# model 5 : SSVM

after we have done an analysis on our dataset, we have found that for some label such as "S-BAR" for chinese address parsing, it is predicted with a very low accuracy. therefore, we have thought of adding a "structual loss" to that kind of label using Structural SVM algorithm. after adding the loss, the model clearly predict the tag with higher loss better, but it hurts the overall model performance. so we did not apply this model in the end.

In [ ]:
AL_para_SSVM_1=AL_para_FOLDER/'SSVM_1'
datafile = AL_train
modelfile = AL_para_SSVM_1
regularization = 10
epoch = 30
ssvm = DM.get_SSVM(datafile, modelfile, regularization)
ssvm.load_data()
ssvm.train(epoch,lr=0.01)

### model performance for AL, since SSVM is only created to deal with imbalanced tagging problem in AL dataset.


Entity  F: 0.8614

Sentiment  F: 0.8535

# Future work

1. found out better and efficient algorithm for implementation of batch training for CRF, probability incorporating stochastic gradient descent.
2. implement latent tree CRF mentioned in this paper: https://www.aclweb.org/anthology/N19-1346/ for chinese address parsing.